File: codes/Deque/C/Deque.c
#include <stdio.h>
#include <stdlib.h>
#include "Deque.h"

void DequeInit(Deque* pdeq)
{
	pdeq->head = NULL;
	pdeq->tail = NULL;
}

int DQIsEmpty(Deque* pdeq)
{
	if (pdeq->head == NULL)
		return TRUE;
	else
		return FALSE;
}

void DQAddFirst(Deque* pdeq, Data data)
{
	Node* newNode = (Node*)malloc(sizeof(Node));
	newNode->data = data;
	newNode->next = NULL;
	newNode->prev = NULL;

	if (DQIsEmpty(pdeq))
	{
		pdeq->tail = newNode;
	}
	else
	{
		newNode->next = pdeq->head;
		pdeq->head->prev = newNode;
	}
	pdeq->head = newNode;
}

void DQAddLast(Deque* pdeq, Data data)
{
	Node* newNode = (Node*)malloc(sizeof(Node));
	newNode->data = data;
	newNode->next = NULL;
	newNode->prev = NULL;

	if (pdeq->tail == NULL)
	{
		pdeq->head = newNode;
	}
	else
	{
		newNode->prev = pdeq->tail;
		pdeq->tail->next = newNode;
	}
	pdeq->tail = newNode;
}

Data DQRemoveFirst(Deque* pdeq)
{
	Node* delNode = pdeq->head;
	Data retData = delNode->data;

	if (DQIsEmpty(pdeq))
	{
		printf("Deque Memory Error!");
		exit(-1);
	}

	if (pdeq->head == pdeq->tail)
	{
		pdeq->head = NULL;
		pdeq->tail = NULL;
	}
	else
	{
		pdeq->head = delNode->next;
		pdeq->head->prev = NULL;
	}

	free(delNode);
	return retData;
}

Data DQRemoveLast(Deque* pdeq)
{
	Node* delNode = pdeq->tail;
	Data retData = delNode->data;

	if (DQIsEmpty(pdeq))
	{
		printf("Deque Memory Error!");
		exit(-1);
	}

	if (pdeq->head == pdeq->tail)
	{
		pdeq->head = NULL;
		pdeq->tail = NULL;
	}
	else
	{
		pdeq->tail = delNode->prev;
		pdeq->tail->next = NULL;
	}

	free(delNode);
	return retData;
}

Data DQGetFirst(Deque* pdeq)
{
	if (DQIsEmpty(pdeq))
	{
		printf("Deque Memory Error!");
		exit(-1);
	}

	return pdeq->head->data;
}

Data DQGetLast(Deque* pdeq)
{
	if (DQIsEmpty(pdeq))
	{
		printf("Deque Memory Error!");
		exit(-1);
	}

	return pdeq->tail->data;
}

File: codes/DequeBaseQueue/Java/DequeBaseQueue/Deque.java
package DequeBaseQueue;

public class Deque<T> implements DequeInterface<T> {
    private Node<T> head;
    private Node<T> tail;

    public Deque() {
        this.head = null;
        this.tail = null;
    }

    @Override
    public boolean DQIsEmpty() {
        if (this.head == null)
            return true;
        else
            return false;
    }

    @Override
    public void DQAddFirst(T data) {
        Node<T> newNode = new Node<T>();
        newNode.setData(data);
        newNode.setNext(null);
        newNode.setPrev(null);

        if (this.DQIsEmpty())
            this.tail = newNode;
        else {
            newNode.setNext(this.head);
            this.head.setPrev(newNode);
        }
        this.head = newNode;
    }

    @Override
    public void DQAddLast(T data) {
        Node<T> newNode = new Node<T>();
        newNode.setData(data);
        newNode.setNext(null);
        newNode.setPrev(null);

        if (this.tail == null)
            this.head = newNode;
        else {
            newNode.setPrev(this.tail);
            this.tail.setNext(newNode);
        }
        this.tail = newNode;
    }

    @Override
    public T DQRemoveFirst() {
        Node<T> delNode = this.head;
        T retData = delNode.getData();

        if (this.DQIsEmpty()) {
            System.out.println("Deque Memory Error!");
            System.exit(-1);
        }

        if (this.head == this.tail) {
            this.head = null;
            this.tail = null;
        } else {
            this.head = delNode.getNext();
            this.head.setPrev(null);
        }

        return retData;
    }

    @Override
    public T DQRemoveLast() {
        Node<T> delNode = this.tail;
        T retData = delNode.getData();

        if (this.DQIsEmpty()) {
            System.out.println("Deque Memory Error!");
            System.exit(-1);
        }

        if (this.head == this.tail) {
            this.head = null;
            this.tail = null;
        } else {
            this.tail = delNode.getPrev();
            this.tail.setNext(null);
        }

        return retData;
    }

    @Override
    public T DQGetFirst() {
        if (this.DQIsEmpty()) {
            System.out.println("Deque Memory Error!");
            System.exit(-1);
        }

        return this.head.getData();
    }

    @Override
    public T DQGetLast() {
        if (this.DQIsEmpty()) {
            System.out.println("Deque Memory Error!");
            System.exit(-1);
        }

        return this.tail.getData();
    }
}
